File: src/zql/parser.zig
const types = @import("../types.zig");
const std = @import("std");
const ctx = @import("ctx.zig");
const lex = @import("lexer.zig");
const err = @import("errors.zig");
const collection = @import("../collection.zig");
const cmds = @import("commands.zig");
const utils = @import("../utils.zig");
const mem = @import("memory");

pub const Parser = struct {
    db: *collection.Collection,
    lexer: lex.Lexer,

    pub fn init(db: *collection.Collection, code: [:0]const u8) Parser {
        return Parser{
            .db = db,
            .lexer = lex.Lexer.init(code),
        };
    }

    pub fn parse(self: *@This(), allocator: std.mem.Allocator) !std.ArrayList(ctx.ZqlContext) {
        var contexts = std.ArrayList(ctx.ZqlContext).init(allocator);
        
        var tok = try self.lexer.parse_token();

        if (tok.type != lex.TokenTypes.identifier) {
            return err.Errors.ExpectedCommandError;
        }

        while (tok.type != lex.TokenTypes.eof) {
            if (cmds.commands.get(tok.text)) |command| {
                var context = ctx.ZqlContext.init(self.db, allocator);

                while (true) {
                    tok = try self.lexer.parse_token();
 
                    if (tok.type == lex.TokenTypes.eof or tok.type == lex.TokenTypes.semicolon) {
                        if (tok.type == lex.TokenTypes.identifier) {
                            try self.lexer.step_back(tok.text.len);
                        }
                        break;
                    }

                    var arg = try self.parse_value(tok, allocator);
                    try context.add_arg(arg);
                }

                context.set_fn(command);

                try contexts.append(context);
            } else {
                var context = ctx.ZqlContext.init(self.db, allocator);
                try context.add_arg(.{ .err = try std.fmt.allocPrint(allocator, "Invalid command or unexpected token at {}:{}!", .{tok.line, tok.col}), .value = null });

                try contexts.append(context);
                return contexts;
            }
            tok = try self.lexer.parse_token();
        }

        return contexts;
    }

    fn parse_primitive_value(self: *@This(), tok: lex.Token, allocator: std.mem.Allocator) anyerror!?ctx.ZqlTrace { 
        switch(tok.type) {
            lex.TokenTypes.identifier, lex.TokenTypes.string => {
                var val = try mem.AutoPtr(types.Value).init(allocator, .{ .String = try utils.strdup(tok.text, allocator) });
                defer val.deinit();
                val.set_on_free(types.dispose);
                return .{ .value = val.move(), .err = null };
            },
            lex.TokenTypes.int => {
                var val = try mem.AutoPtr(types.Value).init(allocator, .{ .Int = try std.fmt.parseInt(i64, tok.text, 10) });
                defer val.deinit();
                val.set_on_free(types.dispose);
                return .{ .value = val.move(), .err = null };
            },
            lex.TokenTypes.float => {
                var val = try mem.AutoPtr(types.Value).init(allocator, .{ .Float = try std.fmt.parseFloat(f64, tok.text) });
                defer val.deinit();
                val.set_on_free(types.dispose);
                return .{ .value = val.move(), .err = null };
            },
            lex.TokenTypes.bool => {
                var bool_val = false;

                if (std.mem.eql(u8, "true", tok.text)) {
                    bool_val = true;
                }

                var val = try mem.AutoPtr(types.Value).init(allocator, .{ .Bool = bool_val });
                defer val.deinit();
                val.set_on_free(types.dispose);

                return .{ .value = val.move(), .err = null };
            },
            lex.TokenTypes.lsquarebracket, lex.TokenTypes.lsquiglybracket => {
                try self.lexer.step_back(1);
                return try self.parse_value(tok, allocator);
            },
            else => {
                return null;
            }
        }
    }

    fn parse_value(self: *@This(), token: lex.Token, allocator: std.mem.Allocator) !ctx.ZqlTrace {
        var tok = token;

        if (tok.type == lex.TokenTypes.eof) {
            return .{ .err = try std.fmt.allocPrint(allocator, "Expected value at {}:{} but got EOF!", .{tok.line, tok.col}), .value = null };
        }

        if (tok.type == lex.TokenTypes.lsquarebracket) {
            var value_array = std.ArrayList(*mem.AutoPtr(types.Value)).init(allocator);

            while (true) {
                tok = try self.lexer.parse_token();
                if (tok.type == lex.TokenTypes.rsquarebracket) {
                    break;
                } else if (tok.type == lex.TokenTypes.comma) {
                    continue;
                } else if (tok.type == lex.TokenTypes.eof) {
                    for (value_array.items) |item| {
                        item.deinit();
                    }
                    value_array.deinit();
                    return .{ .err = try std.fmt.allocPrint(allocator, "Expected , or ] at {}:{} but got EOF!", .{tok.line, tok.col}), .value = null };
                }

                var array_val = try self.parse_primitive_value(tok, allocator);

                if (array_val == null) {
                    for (value_array.items) |item| {
                        item.deinit();
                    }
                    value_array.deinit();
                    return .{ .err = try std.fmt.allocPrint(allocator, "Expected value at {}:{} but got {} instead!", .{tok.line, tok.col, tok.type}), .value = null };
                }
                
                if (array_val.?.err) |_| {
                    return array_val.?;
                }

                if (array_val == null) {
                    for (value_array.items) |item| {
                        item.deinit();
                    }
                    value_array.deinit();

                    return .{ .err = try std.fmt.allocPrint(allocator, "Expected value at {}:{} but got EOF!", .{tok.line, tok.col}), .value = null };
                }

                try value_array.append(array_val.?.value.?);
            }

            var val = try mem.AutoPtr(types.Value).init(allocator, .{ .Array = value_array });
            defer val.deinit();
            val.set_on_free(types.dispose);
            return .{ .value = val.move(), .err = null };
        } else if (tok.type == lex.TokenTypes.lsquiglybracket) {
            var obj = try collection.Collection.init(allocator);

            while (true) {
                tok = try self.lexer.parse_token();
                if (tok.type == lex.TokenTypes.rsquiglybracket) {
                    break;
                }

                if (tok.type != lex.TokenTypes.string and
                    tok.type != lex.TokenTypes.identifier) {
                    obj.deinit(allocator);
                    return .{ .err = try std.fmt.allocPrint(allocator, "Expected String or Identifier but got {} at {}:{}!", .{tok.type, tok.line, tok.col}), .value = null };
                }

                var obj_key = try utils.strdup(tok.text, allocator);
                defer allocator.free(obj_key);

                tok = try self.lexer.parse_token();

                var obj_value: ?ctx.ZqlTrace = null;

                if (tok.type == lex.TokenTypes.colon) {
                    tok = try self.lexer.parse_token();
                    obj_value = try self.parse_primitive_value(tok, allocator);
                }   else {
                    obj_value = try self.parse_primitive_value(tok, allocator);
                }

                if (obj_value == null) {
                    obj.deinit(allocator);
                    return .{ .err = try std.fmt.allocPrint(allocator, "Expected value at {}:{} but got {} instead!", .{tok.line, tok.col, tok.type}), .value = null };
                }

                if (obj_value.?.err) |_| {
                    return obj_value.?;
                }
                try obj.add(obj_key, obj_value.?.value.?, allocator);

                tok = try self.lexer.parse_token();
                if (tok.type != lex.TokenTypes.comma) {
                    try self.lexer.step_back(@intCast(tok.text.len));
                    if (tok.type == lex.TokenTypes.string) {
                        try self.lexer.step_back(2);
                    }
                }
            }

            var val = try mem.AutoPtr(types.Value).init(allocator, .{ .Collection = obj });
            defer val.deinit();
            val.set_on_free(types.dispose);

            return .{ .value = val.move(), .err = null };
        }

        var val = try self.parse_primitive_value(tok, allocator);

        if (val != null) {
            if (val.?.err) |_| {
                return val.?;
            }

            return val.?;
        }

        return .{ .err = try std.fmt.allocPrint(allocator, "Expected value at {}:{} but got {} instead!", .{tok.line, tok.col, tok.type}), .value = null };
    }
};

File: src/zql/parser.cpp
#include <string>
#include <vector>
#include <memory>

#include <logger.hpp>
#include <types.hpp>
#include <collection.hpp>
#include <zql/ctx.hpp>
#include <zql/parser.hpp>

using ZeonDB::Types::Value;

#define EXTRN(N) extern void N(ZeonDB::ZQL::Context *ctx);
#define FN(N) {#N, N}

EXTRN(help);
EXTRN(get);
EXTRN(del);
EXTRN(set);
EXTRN(auth);
EXTRN(template_cmd);
EXTRN(link_cmd);
EXTRN(link_cmd);
EXTRN(branches);
EXTRN(options);
EXTRN(array);

std::map<std::string, ZeonDB::ZQL::ZqlFunction> commands {
	FN(help),
	FN(branches),
	FN(get),
	FN(array),
	{"delete", del},
	FN(set),
	FN(auth),
	{"template", template_cmd},
	{"link", link_cmd},
	FN(options),
};

std::string substr(std::string_view& view, size_t start, size_t len) {
    return std::string({view.begin() + start, view.begin() + start + len});
}

namespace ZeonDB::ZQL {
	ZqlTrace Parser::parse_primitive_value(Token tok) {
		std::shared_ptr<Value> val = nullptr;
		std::string error = "";
		switch (tok.type) {
			case TokenTypes::identifier:
			case TokenTypes::string:
				val = Value::new_string(substr(this->code, tok.col, tok.len));
				break;
			case TokenTypes::integer:
				val = Value::new_int(std::stoi(substr(this->code, tok.col, tok.len)));
				break;
			case TokenTypes::floating:
				val = Value::new_float(std::stof(substr(this->code, tok.col, tok.len)));
				break;
			case TokenTypes::boolean:
			{
				bool bool_val = false;

				std::string str = substr(this->code, tok.col, tok.len);
				if (str.compare("true") == 0) {
					bool_val = true;
				}

				val = Value::new_bool(bool_val);
				break;
			}
			case TokenTypes::lsquarebracket:
			case TokenTypes::lsquiglybracket:
			{
				ZqlTrace trace = this->parse_value(tok);
				if (trace.value)
					val = trace.value;
				else if (trace.error.length() > 0)
					error = trace.error;
				break;
			}
			default:
				val = nullptr;
				error = "Invalid token at parse_primitive_value " + std::to_string(tok.line) + ":" + std::to_string(tok.col);
		}

		return (ZqlTrace) {
			.value = val,
			.error = error
		};
	}

	ZqlTrace Parser::parse_value(Token tok) {
		if (tok.type == TokenTypes::eof) {
			return (ZqlTrace) {
				.value = nullptr,
				.error = "Expected value at " + std::to_string(tok.line) + ":" + std::to_string(tok.col) + " but got EOF!",
			};
		}

		if (tok.type == TokenTypes::lsquarebracket) {
			LOG_D("Parsing array", nullptr);
			std::vector<std::shared_ptr<Value>> values;

			while (true) {
				tok = this->lexer.parse_token();
				if (tok.type == TokenTypes::rsquarebracket) break;
				else if (tok.type == TokenTypes::comma) continue;
				else if (tok.type == TokenTypes::eof)
					return (ZqlTrace) {
						.value = nullptr,
						.error = "Expected , or ] at " + std::to_string(tok.line) + ":" + std::to_string(tok.col)
					};

				auto val = this->parse_primitive_value(tok);

				if (val.error.length() > 0) {
					return val;
				}

				values.push_back(val.value);
			}

			auto array = Value::new_array();
			array->v.a = values;
			return (ZqlTrace) {
				.value = array,
				.error = ""
			};
		} else if (tok.type == TokenTypes::lsquiglybracket) {
			LOG_D("Parsing collection", nullptr);
			Collection object;

			bool key_parsed = false;
			std::string key;

			while (true) {
				tok = this->lexer.parse_token();
				if (key_parsed == false) {
					if (tok.type == TokenTypes::rsquiglybracket) break;
					else if (tok.type == TokenTypes::comma) continue;
					else if (tok.type == TokenTypes::eof)
						return (ZqlTrace) {
							.value = nullptr,
							.error = "Expected , or } at " + std::to_string(tok.line) + ":" + std::to_string(tok.col)
						};
				} else {
					if (tok.type == TokenTypes::colon) continue;
					else if (tok.type == TokenTypes::rsquiglybracket || tok.type == TokenTypes::eof)
						return (ZqlTrace) {
							.value = nullptr,
							.error = "Expected value at " + std::to_string(tok.line) + ":" + std::to_string(tok.col)
						};
				}

				if (key_parsed == false) {
					key = this->code.substr(tok.col, tok.len);
					key_parsed = true;
					continue;
				}

				auto val = this->parse_primitive_value(tok);

				if (val.error.length() > 0) {
					return val;
				}

				object.add(key, val.value);
				key_parsed = false;
			}

			auto obj = Value::new_collection();
			obj->v.c = object;
			return (ZqlTrace) {
				.value = obj,
				.error = ""
			};
		}

		return this->parse_primitive_value(tok);
	}

	std::vector<Context> Parser::parse() {
		std::vector<Context> ctxs;

		Token tok = this->lexer.parse_token();

		LOG_D("Got token %s %d", substr(this->code, tok.col, tok.len).c_str(), tok.type);
		if (tok.type != TokenTypes::identifier) {
			Context ctx(this->db, "Expected identifier at " + std::to_string(tok.line) + ":" + std::to_string(tok.col));
			ctxs.push_back(ctx);
			return ctxs;
		}

		while (tok.type != TokenTypes::eof && tok.type != TokenTypes::illegal) {
			if (tok.type != TokenTypes::identifier) {
				Context ctx(this->db, "Expected identifier at " + std::to_string(tok.line) + ":" + std::to_string(tok.col));
				ctxs.push_back(ctx);
				return ctxs;
			} else {
				Context ctx(this->db);
				std::string fn_name = substr(this->code, tok.col, tok.len);
				if (!commands.contains(fn_name)) {
					Context ctx(this->db, "Unknown command at " + std::to_string(tok.line) + ":" + std::to_string(tok.col));
					ctxs.push_back(ctx);
					return ctxs;
				}

				ctx.set_function(commands[fn_name]);
				while (true) {
					tok = this->lexer.parse_token();
					if (tok.type == TokenTypes::eof or tok.type == TokenTypes::semicolon) break;

					LOG_D("Parsing token %s", substr(this->code, tok.col, tok.len).c_str());
					ZqlTrace arg = this->parse_value(tok);
					ctx.add_arg(arg);
					if (arg.error.length() > 0) {
						LOG_D("Context recieved an error!", nullptr);
						ctx.error = arg.error;
						break;
					}
				}

				LOG_D("Context parsed successfully!", nullptr);
				ctxs.push_back(ctx);
			}
			tok = this->lexer.parse_token();
		}
		if (tok.type == TokenTypes::illegal) {
			Context ctx(this->db, "Invalid token");
			ctxs.push_back(ctx);
		}

		return ctxs;
	}
}
